schema/invalid_serialization_detector.go

```diff
@@ -0,0 +1,83 @@
+package schema
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// NewInvalidSerializationDetectorSchema creates a new test schema type.
+func NewInvalidSerializationDetectorSchema() *InvalidSerializationDetectorSchema {
+	return &InvalidSerializationDetectorSchema{}
+}
+
+// InvalidSerializationDetectorSchema is a testing type that detects double serialization or double unserialization,
+// which helps prevent bugs.
+// It is backed by a string, which is set to the last operation performed. If the last operation matches the
+// current operation, then that indicates double serialization or double unserialization.
+type InvalidSerializationDetectorSchema struct{}
+
+func (d InvalidSerializationDetectorSchema) Unserialize(data any) (any, error) {
+	// The input is expected to always be a string.
+	// If the input is equal to "unserialized", then that means it's being unserialized a second time.
+	return d.detectInvalidValue("unserialized", data)
+}
+
+func (d InvalidSerializationDetectorSchema) detectInvalidValue(operation string, data any) (any, error) {
+	asString, isString := data.(string)
+	if !isString {
+		return nil, &ConstraintError{
+			Message: fmt.Sprintf(
+				"unsupported data type for InvalidSerializationDetectorSchema; expected string, got %T",
+				data,
+			),
+		}
+	}
+	if asString == operation {
+		return nil, &ConstraintError{
+			Message: fmt.Sprintf("InvalidSerializationDetectorSchema double %s", operation),
+		}
+	}
+	return operation, nil
+}
+
+func (d InvalidSerializationDetectorSchema) UnserializeType(data any) (string, error) {
+	unserialized, err := d.Unserialize(data)
+	if err != nil {
+		return "", err
+	}
+	return unserialized.(string), nil
+}
+
+func (d InvalidSerializationDetectorSchema) ValidateCompatibility(_ any) error {
+	// Not applicable to this data type
+	return nil
```

Comment:
If this function is not applicable, why does it return "no-error"?  It seems like it should either return an error (claiming that it is not applicable and/or that it should not have been called) or it should panic.

Reply:
So, the comment that this function is "not applicable to this data type" is not accurate...instead, it's more the case either that this data type is never incompatible or that we just don't care...and, regardless, the action here is to never return an error when the check is made.